storage/repository.go

```diff
@@ -0,0 +1,152 @@
+// would not be usable in a useful way. Some additional methods which can offer
+// more optimization opportunities can also be implemented by calling the
+// appropriate Register...() function inside of the Repository.Init() function.
+// For example, storage.Copy() will by default use a reader and a writer to
+// perform a copy operation, but a repository might use RegisterCopy() to cause
+// a more optimal version to be used for URIs of the relevant scheme (e.g. by
+// performing the copy on the remote side of an ssh connection).
+//
+// In some cases, you may wish to omit an implementation of one of these
+// functions anyway. In such cases, you should do this by returning a
+// URIOperationNotSupportedError. For example, this may be appropriate when
+// building a storage repository which represents only resources that are
+// read-only - in such cases functions like WriterTo() may not make sense.
+// Please use care when deciding to do this, and consider how it will impact
+// users of your storage repository.
+//
+// NOTE: most developers who use Fyne should *not* generally attempt to
+// call repository methods directly. You should use the methods in the storage
+// package, which will automatically detect the scheme of a URI and call into
+// the appropriate repository.
+//
+// NOTE: functions in a particular repository can assume they will only ever be
+// used on URIs which match the scheme that they have been registered to
+// handle, however they should not assume that they will only be called on URIs
+// of the same implementation as the ParseURI() for this repository has
+// returned.
+//
+// Since: 2.0.0
+type Repository interface {
+
+	// ParseURI returns a method that given a text string creates a
+	// new URI instance. This may cause an error, for example if the string
+	// is an invalid URI or contains information that is invalid in the
+	// context of the repository's scheme.
+	//
+	// This method will be called only after the URI scheme has been
+	// verified to match the one the repository was registered to handle,
+	// however the URI is not validated further than this. It is required
+	// by convention (but not enforced by technical means) that this
+	// function should throw an error if the string is not a valid IETF RFC
+	// 3986 complaint URI. Fail to validate this at your own risk.
+	//
+	// NOTE: it is highly recommended to use net/url (
+	// https://golang.org/pkg/net/url/ ) for parsing URI strings unless you
+	// have a good reason not to. It is mature and implements RFC 3986.
+	//
+	// If no implementation is provided, a generic URI based on RFC3986 is
+	// parsed without any special validation logic.
+	//
+	// Since 2.0.0
+	ParseURI() func(string) (fyne.URI, error)
+
+	// Init will be called while the repository is being registered.  This
+	// is the appropriate place to register any optional handlers relevant
+	// to a particular URI scheme.
+	//
+	// Since 2.0.0
+	Init() func() error
```

Comment:
I'm not sure about including `Init` to be run during register - if we are supporting a single `Repository` backing multiple schemes then this could be called multiple times, which the name does not imply.
What is this needed for that cannot be expected of the implementation before calling `Register`?

Reply:
I am also not sure this needs to be included. I envisioned the initialization going the other way, where the developer initializes a repository, most likely by importing a third party library, which runs its `func init()`, which registers itself with the storage API (like when importing "image/png")